README.md
# event-stream-client

[![Node.js CI](https://github.com/TREEcg/event-stream-client/actions/workflows/ci.yml/badge.svg)](https://github.com/TREEcg/event-stream-client/actions/workflows/ci.yml) [![npm version](https://img.shields.io/npm/v/@treecg/actor-init-ldes-client.svg?style=popout)](https://www.npmjs.com/package/@treecg/actor-init-ldes-client)

## Install

```bash
git clone https://github.com/TREEcg/event-stream-client.git
cd event-stream-client
npm install
```

## Sync

If you want to sync with an [Linked Data Event Stream (LDES)](https://w3id.org/ldes/specification), have a look at the pre-built engine:

- [actor-init-ldes-client](https://github.com/TREEcg/event-stream-client/tree/main/packages/actor-init-ldes-client): Comunica-based engine to materialize sync with an LDES from JavaScript applications or the CLI.

## Publish

```bash
npm run publish
```


CHANGELOG.md
# Change Log

<a name="v4.0.8"></a>
## [v4.0.8](https://github.com/TREEcg/event-stream-client/compare/v4.0.7...v4.0.8) - 2023-05-17

### TODO: categorize commits, choose titles from: Added, Changed, Deprecated, Removed, Fixed, Security.
* [Remove unused dependencies on components.js config](https://github.com/TREEcg/event-stream-client/commit/a80add3551f6b82480cc59d752f326275b2aea4d)
* [Add NPM badge to README](https://github.com/TREEcg/event-stream-client/commit/c18b8320abd6a64589adc4f6f2b8e9a8528e5920)
* [Bump version](https://github.com/TREEcg/event-stream-client/commit/68d24edffa0399333d35b31bab1617a458dfef6a)

<a name="v4.0.3"></a>
## [v4.0.3](https://github.com/TREEcg/event-stream-client/compare/v4.0.2...v4.0.3) - 2023-03-29

### TODO: categorize commits, choose titles from: Added, Changed, Deprecated, Removed, Fixed, Security.
* [Move @treecg/types to runtime dependencies](https://github.com/TREEcg/event-stream-client/commit/e671880056c3d10f7affab8f74bbd55c57f03319)

<a name="v4.0.2"></a>
## [v4.0.2](https://github.com/TREEcg/event-stream-client/compare/v4.0.1...v4.0.2) - 2023-03-29

### TODO: categorize commits, choose titles from: Added, Changed, Deprecated, Removed, Fixed, Security.
* [Expose all types from EventStream](https://github.com/TREEcg/event-stream-client/commit/1e09c49b22a2d4771c2f8ad6b1e22ae7212020b5)
* [bump git head](https://github.com/TREEcg/event-stream-client/commit/2f97dee9bc73f0bc0eb40221b4f789165007ec04)

<a name="v4.0.1"></a>
## [v4.0.1](https://github.com/TREEcg/event-stream-client/compare/v4.0.0...v4.0.1) - 2023-03-28

### TODO: categorize commits, choose titles from: Added, Changed, Deprecated, Removed, Fixed, Security.
* [Update components.js config](https://github.com/TREEcg/event-stream-client/commit/097cbb5d2721d825b36d31d276c237a03563cd3b)

<a name="v4.0.0"></a>
## [v4.0.0](https://github.com/TREEcg/event-stream-client/compare/v3.0.2...v4.0.0) - 2023-03-28

### TODO: categorize commits, choose titles from: Added, Changed, Deprecated, Removed, Fixed, Security.
* [Merge pull request #66 from TREEcg/fix/update-dependencies](https://github.com/TREEcg/event-stream-client/commit/e65ea920e74ac24e281ae1aff5f321b542f13629)
* [Remove coverage threshold (for now)](https://github.com/TREEcg/event-stream-client/commit/34234fe9e2323374381e7366c5708158b864ec71)
* [Fix component.js build process](https://github.com/TREEcg/event-stream-client/commit/6e679ca7df3a45c3ffa904ff9385c40278f11ebe)
* [Go back to npm ci + fully remove yarn commands](https://github.com/TREEcg/event-stream-client/commit/df6e6e6715de803884e1318e570ee2c09e866d29)
* [Try normal npm i](https://github.com/TREEcg/event-stream-client/commit/f77f78037a1ecba31472a8879b628dcf4ac4beba)
* [Commit package-lock.json](https://github.com/TREEcg/event-stream-client/commit/e7077881acfc4a93cbe55925d9ad051d7f8e7030)
* [Add github action for CI](https://github.com/TREEcg/event-stream-client/commit/acd58da4c5273ce28654dedb049c82190d26ffa3)
* [Changes include: - Fix time-based prunning of members - Extract ldes:timestampPath from LDES metadata - Handle edge case where no members are emitted on sync mode](https://github.com/TREEcg/event-stream-client/commit/855010eec1d365608d70124e88cf45051719dfc7)
* [Changes include: - Remove unmaintained and stale dependencies - Use WebStreams to process fetched pages - Use RDFJS store to do BGP lookups - Clean up code](https://github.com/TREEcg/event-stream-client/commit/3c07df3e7e4d9bab6f90826b81eab9381655ec44)
* [Remove in-memory caching of all page requests](https://github.com/TREEcg/event-stream-client/commit/453635482b4e5b94009d058838dc3c921b4048f8)
* [Fix cli error message](https://github.com/TREEcg/event-stream-client/commit/54c20fd5a062f79beae90192de380d47ab49e353)
* [Changes include: - Define enum type for output representation - Rename sychronization event - Make sure Event Stream tests run properly](https://github.com/TREEcg/event-stream-client/commit/a79270d5ec8374772f3a64ff642c275de7ec3e8b)
* [Fix LDES client unit tests + properly close stream](https://github.com/TREEcg/event-stream-client/commit/1ffd6b928c3a11a773f5c53ad7a615bc85b11cea)
* [Fix library unit tests + issue on stream pause/flow modes](https://github.com/TREEcg/event-stream-client/commit/c33a7097fa52ada8d514f66e26457b8dcd4142e3)
* [Fix bookkeeper unit tests](https://github.com/TREEcg/event-stream-client/commit/218501fbbefe9b1145d24bf969f7ae386bde3039)
* [Clean up dependencies](https://github.com/TREEcg/event-stream-client/commit/5bc95cb126c13a3cb32394db71875321aada22eb)
* [Update dependencies](https://github.com/TREEcg/event-stream-client/commit/0dd2114766b594cdf51781145dd5d6a8fa70577d)
* [Fix unit tests of RDF object filter with a quadstore](https://github.com/TREEcg/event-stream-client/commit/2b751ddb41bc758eb7d3d1554a768c4156374d13)
* [Fix unit tests of rdf filtering with json-ld framing actor](https://github.com/TREEcg/event-stream-client/commit/fb2e98b983737cf9bf0020ef4785a8e93bdcd16b)
* [Clean up and update more dependencies](https://github.com/TREEcg/event-stream-client/commit/60f3c20a42f2de7a2ca0679c31b83b65fbfca56f)
* [Update jest and jsonld dependencies](https://github.com/TREEcg/event-stream-client/commit/e44a6367081acccfcf1600947c0a9caf6d851610)
* [Update lerna and work with npm](https://github.com/TREEcg/event-stream-client/commit/9ce2a8e0c546a481964f0851c1e3f626ba9efd2f)
* [Update typescript dependencies](https://github.com/TREEcg/event-stream-client/commit/26cfa56f61d691a6348120bd07c91d3c38f132b9)

<a name="v3.0.2"></a>
## [v3.0.2](https://github.com/TREEcg/event-stream-client/compare/v3.0.1...v3.0.2) - 2023-03-03

### TODO: categorize commits, choose titles from: Added, Changed, Deprecated, Removed, Fixed, Security.
* [Fix logic to allow for readable stream paused mode](https://github.com/TREEcg/event-stream-client/commit/211c03666c937924b3ad02eec2ec87c48d0b825b)

test/example.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var Sync = require('../dist/sync.js').Sync;
var sync = new Sync();
var main = function () {
    return __awaiter(this, void 0, void 0, function () {
        var url, options, eventstreamSync;
        return __generator(this, function (_a) {
            try {
                url = "http://lodi.ilabt.imec.be/coghent/objecten";
                options = {
                    "pollingInterval": 5000 // millis
                };
                eventstreamSync = sync.createReadStream(url, options);
                eventstreamSync.on('data', function (data) {
                    console.log(data);
                });
            }
            catch (e) {
                console.error(e);
            }
            return [2 /*return*/];
        });
    });
};
main();


lib/sync.js
const FETCH_PAUSE = 2000; // in milliseconds; pause before fetching the next fragment
let pollingInterval = 10000; // in milliseconds; when the response is cacheable, wait this time before refetching it

const followRedirects = require('follow-redirects');
followRedirects.maxRedirects = 10;
const { http, https } = followRedirects;
const CacheableRequest = require('cacheable-request');
const cacheableRequestHttp = new CacheableRequest(http.request);
const cacheableRequestHttps = new CacheableRequest(https.request);
const CachePolicy = require('http-cache-semantics');

const rdfParser = require("rdf-parse").default;
const N3 = require('n3');
const streamWriter = new N3.StreamWriter({ prefixes: { tree: 'https://w3id.org/tree#' } });

const Bookkeeper = require('./bookkeeper');
let bk = new Bookkeeper();

function createReadStream (url, options) {
    if (!url) {
        console.error('Provide a URI of a TREE root node please');
        process.exit();
    }
    if (options && options.pollingInterval) pollingInterval = options.pollingInterval;

    retrieve(url);

    return streamWriter;
}

async function retrieve(pageUrl) {
    console.error('GET ' + pageUrl);
    let startTime = new Date();
    try {
        let page = await getPage(pageUrl);
        let endTime = new Date();
        let isFromCache = page.fromCache;

        if (isFromCache) {
            // this should never happen, because timeToLive should be expired
            console.error('' + page.statusCode + ' ' + pageUrl + ' (' + (endTime.getTime() - startTime.getTime()) + 'ms)');
            // Add again to book keeper
            // use pageUrl, because page.url is undefined when coming from client-side cache
            bk.addFragment(pageUrl, pollingInterval);
            retrieveRecursively();
        } else {
            console.error('' + page.statusCode + ' ' + page.url + ' (' + (endTime.getTime() - startTime.getTime()) + 'ms)');
            const policy = new CachePolicy(page.request, page.response);
            const ttl = policy.storable() ? policy.timeToLive() : pollingInterval;
            bk.addFragment(page.url, ttl);

            let dataStream = require('streamify-string')(page.data.toString());
            // only keep media-type directive of content-type header
            let mediaType = page.headers['content-type'].indexOf(';') > 0 ? page.headers['content-type'].substr(0, page.headers['content-type'].indexOf(';')) : page.headers['content-type'];
            rdfParser.parse(dataStream, {contentType: mediaType})
                .on('data', (quad) => {
                    streamWriter.write(quad);
                    if (quad.predicate.value === "https://w3id.org/tree#node") {
                        // later: add relation metadata so the bookkeeper can prune
                        bk.addFragment(quad.object.value, 0);
                    }
                })
                .on('error', (error) => console.error(error))
                .on('end', () => retrieveRecursively());
        }
    } catch (e) {
        console.error('Failed to retrieve ' + pageUrl + ': ' + e);
        retrieveRecursively();
    }
}

async function retrieveRecursively() {
    await sleep(FETCH_PAUSE);
    if (bk.nextFragmentExists()) {
        let next = bk.getNextFragmentToFetch();
        let now = new Date();
        // Do not refetch too soon
        while (next.refetchTime.getTime() > now.getTime()) {
            await sleep(FETCH_PAUSE);
            now = new Date();
        }
        retrieve(next.url);
    } else {
        // We're done
        streamWriter.push(null);
    }
}

function getPage(pageUrl) {
    return new Promise((resolve, reject) => {
        const protocol = new URL(pageUrl).protocol;
        let r = protocol === 'https:' ? cacheableRequestHttps : cacheableRequestHttp;

        const cacheReq = r(pageUrl, (res) => {
            let data = '';

            // A chunk of data has been recieved.
            res.on('data', (chunk) => {
                data += chunk;
            });

            // The whole response has been received. Print out the result.
            res.on('end', () => {
                // This is necessary for cachePolicy
                res.req.headers = res.headers;
                resolve({
                    "request": res.req,
                    "response": res,
                    "headers": res.headers,
                    "url": res.responseUrl,
                    "data": data,
                    "statusCode": res.statusCode,
                    "fromCache": res.fromCache
                });
            })
        })

        cacheReq.on('request', request => request.end());
        cacheReq.on('error', (e) => reject(e));
    });
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

module.exports = sync = { createReadStream };


lib/bookkeeper.js
const PriorityQueue = require('js-priority-queue');

class Bookkeeper {
    constructor(options) {
        this.queue = new PriorityQueue({ comparator: compareFragments});
        this.queued = {}; // to know whether a fragment URL is already added to the priority queue
    }
}

Bookkeeper.prototype.addFragment = function (url, ttl) {
    if (!Object.keys(this.queued).includes(url)) {
        let fragmentInfo = {
            "url": url,
            "refetchTime": new Date(new Date().getTime() + ttl) // now
        }
        this.queue.queue(fragmentInfo);
        this.queued[url] = {};
    }
}

function compareFragments(a, b) {
    return a.refetchTime.getTime() - b.refetchTime.getTime();
};

Bookkeeper.prototype.nextFragmentExists = function() {
    return this.queue.length > 0 ? true : false;
}

Bookkeeper.prototype.getNextFragmentToFetch = function() {
    let next = this.queue.dequeue();
    delete this.queued[next.url];
    return next;
}

module.exports = Bookkeeper;

lib/bookkeeper.ts
const PriorityQueue = require('js-priority-queue');

export class Bookkeeper {
    protected readonly queue = new PriorityQueue({ comparator: compareFragments});
    protected readonly queued: any = {}; // to know whether a fragment URL is already added to the priority queue

    public constructor() {

    }

    public addFragment(url: string, ttl: number) {
        if (!Object.keys(this.queued).includes(url)) {
            let fragmentInfo = {
                "url": url,
                "refetchTime": new Date(new Date().getTime() + ttl) // now
            }
            this.queue.queue(fragmentInfo);
            this.queued[url] = {};
        }
    }

    public nextFragmentExists() {
        return this.queue.length > 0 ? true : false;
    }

    public getNextFragmentToFetch() {
        let next = this.queue.dequeue();
        delete this.queued[next.url];
        return next;
    }
}


function compareFragments(a: any, b: any) {
    return a.refetchTime.getTime() - b.refetchTime.getTime();
};

lib/sync.ts
const FETCH_PAUSE = 2000; // in milliseconds; pause before fetching the next fragment
let pollingInterval = 10000; // in milliseconds; when the response is cacheable, wait this time before refetching it

const followRedirects = require('follow-redirects');
followRedirects.maxRedirects = 10;
const { http, https } = followRedirects;
const CacheableRequest = require('cacheable-request');
const cacheableRequestHttp = new CacheableRequest(http.request);
const cacheableRequestHttps = new CacheableRequest(https.request);
const CachePolicy = require('http-cache-semantics');

import rdfParser from "rdf-parse";
import * as RDF from "rdf-js";
const N3 = require('n3');
const streamWriter = new N3.StreamWriter({ prefixes: { tree: 'https://w3id.org/tree#' } });
const RdfObjectLoader = require("rdf-object").RdfObjectLoader;
const JsonLdSerializer = require("jsonld-streaming-serializer").JsonLdSerializer;
const jsonLdSerializer = new JsonLdSerializer({ space: '  ', context: [
        "https://data.vlaanderen.be/doc/applicatieprofiel/cultureel-erfgoed-object/kandidaatstandaard/2020-07-17/context/cultureel-erfgoed-object-ap.jsonld",
        "https://data.vlaanderen.be/context/persoon-basis.jsonld",
        "https://brechtvdv.github.io/demo-data/cultureel-erfgoed-event-ap.jsonld",
        {
            "dcterms:isVersionOf": {
"@type": "@id"
},
"prov": "http://www.w3.org/ns/prov#"
}
] });
const {namedNode, literal, triple} = require("@rdfjs/data-model");

//const Bookkeeper = require('./bookkeeper');
import { Bookkeeper } from './bookkeeper';
let bk = new Bookkeeper();

export class Sync {
    public constructor() {
    }

    public createReadStream(url: string, options: { pollingInterval: number; }) {
        if (!url) {
            console.error('Provide a URI of a TREE root node please');
            process.exit();
        }
        if (options && options.pollingInterval) pollingInterval = options.pollingInterval;

        this.retrieve(url);
        return jsonLdSerializer;
        // return streamWriter;
    }

    public async retrieve(pageUrl: string) {
        console.error('GET ' + pageUrl);
        let startTime = new Date();
        try {
            let page = await this.getPage(pageUrl);
            let endTime = new Date();
            let isFromCache = page.fromCache;

            if (isFromCache) {
                // this should never happen, because timeToLive should be expired
                console.error('' + page.statusCode + ' ' + pageUrl + ' (' + (endTime.getTime() - startTime.getTime()) + 'ms)');
                // Add again to book keeper
                // use pageUrl, because page.url is undefined when coming from client-side cache
                bk.addFragment(pageUrl, pollingInterval);
                this.retrieveRecursively();
            } else {
                console.error('' + page.statusCode + ' ' + page.url + ' (' + (endTime.getTime() - startTime.getTime()) + 'ms)');
                const policy = new CachePolicy(page.request, page.response);
                const ttl = policy.storable() ? policy.timeToLive() : pollingInterval;
                bk.addFragment(page.url, ttl);

                let dataStream = require('streamify-string')(page.data.toString());
                // only keep media-type directive of content-type header
                let mediaType = page.headers['content-type'].indexOf(';') > 0 ? page.headers['content-type'].substr(0, page.headers['content-type'].indexOf(';')) : page.headers['content-type'];
                let collectionURI;
                rdfParser.parse(dataStream, {contentType: mediaType})
                    .on('data', (quad: RDF.Quad) => {
                        // streamWriter.write(quad);
                        jsonLdSerializer.write(quad);
                        if (quad.object.value === 'https://w3id.org/tree#Collection') collectionURI = quad.subject.value;
                        if (quad.predicate.value === "https://w3id.org/tree#node") {
                            // later: add relation metadata so the bookkeeper can prune
                            bk.addFragment(quad.object.value, 0);
                        }
                    })
                    .on('error', (error: any) => console.error(error))
                    .on('end', () => {
                        // now we know the collectionURI
                        //getMembers(collectionURI, rdfParser.parse(require('streamify-string')(page.data.toString()), {contentType: mediaType}));
                        this.retrieveRecursively()
                    });
            }
        } catch (e) {
            console.error('Failed to retrieve ' + pageUrl + ': ' + e);
            this.retrieveRecursively();
        }
    }

    /*async function getMembers(collection, parsedStream) {
        const myLoader = new RdfObjectLoader({ context: "https://data.vlaanderen.be/doc/applicatieprofiel/cultureel-erfgoed-object/kandidaatstandaard/2020-07-17/context/cultureel-erfgoed-object-ap.jsonld" }) ;
        myLoader.import(parsedStream).then(() => {
            let members = myLoader.resources[collection].properties['https://w3id.org/tree#member'];
            for (let m in members) {
                // we need to retrieve all quads that are linked with this member
                let member = members[m];
                for (let property in member.properties) {
                    let p = member.properties[property];
                    console.log(property)

                }
                // output quads
                //streamWriter.write(output);
            }

            // Get property values by shortcut
            const myResource = myLoader.resources[collection];
            console.log(`URI:  ${myResource.value}`);
            console.log(`Term type: ${myResource.type}`);
            console.log(`Term value: ${myResource.value}`);
            console.log(`Term: ${myResource.term}`);
        });
    }*/

    public async retrieveRecursively() {
        await this.sleep(FETCH_PAUSE);
        if (bk.nextFragmentExists()) {
            let next = bk.getNextFragmentToFetch();
            let now = new Date();
            // Do not refetch too soon
            while (next.refetchTime.getTime() > now.getTime()) {
                await this.sleep(FETCH_PAUSE);
                now = new Date();
            }
            this.retrieve(next.url);
        } else {
            // We're done
            streamWriter.push(null);
        }
    }

    public getPage(pageUrl: string): Promise<PageMetadata> {
        return new Promise((resolve, reject) => {
            const protocol = new URL(pageUrl).protocol;
            let r = protocol === 'https:' ? cacheableRequestHttps : cacheableRequestHttp;

            const cacheReq = r(pageUrl, (res: any) => {
                let data = '';

                // A chunk of data has been recieved.
                res.on('data', (chunk: any) => {
                    data += chunk;
                });

                // The whole response has been received. Print out the result.
                res.on('end', () => {
                    // This is necessary for cachePolicy
                    res.req.headers = res.headers;
                    resolve({
                        "request": res.req,
                        "response": res,
                        "headers": res.headers,
                        "url": res.responseUrl,
                        "data": data,
                        "statusCode": res.statusCode,
                        "fromCache": res.fromCache
                    });
                })
            })

            cacheReq.on('request', (request : any) => request.end());
            cacheReq.on('error', (e: any) => reject(e));
        });
    }

    public sleep(ms: number) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

class PageMetadata {
    public "request": object;
    public "response": object;
    public "headers": any;
    public "url": any;
    public "data": string;
    public "statusCode": number;
    public "fromCache": boolean;
}